File: EEG app Android/app/src/main/java/com/tutsplus/bleadvertising/DataLoader.java
package com.tutsplus.bleadvertising;

import android.content.Context;

import java.io.IOException;
import java.io.InputStream;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;

public class DataLoader {

    /**
     * Expects a binary file in assets: a 16-byte header + float data + optional label data.
     * - header (little-endian):
     *    int numArrays
     *    int dim1
     *    int dim2
     *    int labelsPresent (1 or 0)
     * - Next: numArrays * dim1 * dim2 floats (4 bytes each)
     * - If labelsPresent == 1, then numArrays * 4 bytes for int labels.
     *
     * Example for 1024 samples, each 18 floats:
     *    numArrays = 1024
     *    dim1 = 18
     *    dim2 = 1   (or 1 if it’s just 1 row)
     *    labelsPresent = 1 or 0
     */
    public DataSample[] loadDataAndLabels(Context context, String assetFileName) throws IOException {
        InputStream inputStream = context.getAssets().open(assetFileName);

        // 4 integers in header => 4 * 4 = 16 bytes
        int headerSize = 4 * 4;
        byte[] headerBytes = new byte[headerSize];
        inputStream.read(headerBytes);

        ByteBuffer headerBuffer = ByteBuffer.wrap(headerBytes).order(ByteOrder.LITTLE_ENDIAN);

        int numArrays     = headerBuffer.getInt();  // number of DataSample
        int dim1          = headerBuffer.getInt();  // first dimension
        int dim2          = headerBuffer.getInt();  // second dimension
        int labelsPresent = headerBuffer.getInt();  // 1 means labels exist

        int totalFloats = numArrays * dim1 * dim2;

        // Read float data using helper method
        int dataSize = totalFloats * 4; // float32 => 4 bytes
        byte[] dataBytes = readBytesFromStream(inputStream, dataSize);
        ByteBuffer dataBuffer = ByteBuffer.wrap(dataBytes).order(ByteOrder.LITTLE_ENDIAN);

        // If labels are present, read them using helper method
        int[] labels = new int[numArrays];
        if (labelsPresent == 1) {
            int labelsSize = numArrays * 4;
            byte[] labelsBytes = readBytesFromStream(inputStream, labelsSize);
            ByteBuffer labelsBuffer = ByteBuffer.wrap(labelsBytes).order(ByteOrder.LITTLE_ENDIAN);
            for (int i = 0; i < numArrays; i++) {
                labels[i] = labelsBuffer.getInt();
            }
        } else {
            // if no labels, just set them to -1
            for (int i = 0; i < numArrays; i++) {
                labels[i] = -1;
            }
        }

        inputStream.close();

        // Build the DataSample array
        DataSample[] dataSamples = new DataSample[numArrays];
        for (int i = 0; i < numArrays; i++) {
            float[] sampleData = new float[dim1 * dim2];
            for (int j = 0; j < (dim1 * dim2); j++) {
                sampleData[j] = dataBuffer.getFloat();
            }
            dataSamples[i] = new DataSample(sampleData, labels[i]);
        }
        return dataSamples;
    }

    /**
     * Helper method to read a specific number of bytes from an InputStream.
     * Handles the case where read() might not return all bytes in one call.
     *
     * @param inputStream The input stream to read from
     * @param size The number of bytes to read
     * @return byte array containing the read bytes
     * @throws IOException if an I/O error occurs
     */
    private byte[] readBytesFromStream(InputStream inputStream, int size) throws IOException {
        byte[] bytes = new byte[size];
        int bytesRead = 0;
        while (bytesRead < size) {
            int result = inputStream.read(bytes, bytesRead, size - bytesRead);
            if (result == -1) break;
            bytesRead += result;
        }
        return bytes;
    }
}

File: APP/app/src/main/java/com/epfl/ch/seizureguard/dl/DataLoader.kt
package com.epfl.ch.seizureguard.dl

import android.content.Context
import android.os.Parcel
import android.os.Parcelable
import java.nio.ByteBuffer
import java.nio.ByteOrder

data class DataSample(
    val data: FloatArray,
    val label: Int
) : Parcelable {
    constructor(parcel: Parcel) : this(
        parcel.createFloatArray()!!,
        parcel.readInt()
    )

    override fun writeToParcel(parcel: Parcel, flags: Int) {
        parcel.writeFloatArray(data)
        parcel.writeInt(label)
    }

    override fun describeContents(): Int = 0
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as DataSample

        if (!data.contentEquals(other.data)) return false
        if (label != other.label) return false

        return true
    }

    override fun hashCode(): Int {
        var result = data.contentHashCode()
        result = 31 * result + label
        return result
    }

    companion object CREATOR : Parcelable.Creator<DataSample> {
        override fun createFromParcel(parcel: Parcel): DataSample {
            return DataSample(parcel)
        }

        override fun newArray(size: Int): Array<DataSample?> {
            return arrayOfNulls(size)
        }
    }
}

class DataLoader {
    fun loadDataAndLabels(context: Context, name: String): Array<DataSample> {
        val assetManager = context.assets
        val inputStream = assetManager.open(name)

        // Read header (4 integers)
        val headerSize = 4 * 4  // 4 integers * 4 bytes per int32
        val headerBytes = ByteArray(headerSize)
        inputStream.read(headerBytes)
        val headerBuffer = ByteBuffer.wrap(headerBytes).order(ByteOrder.LITTLE_ENDIAN)
        val numArrays = headerBuffer.int
        val dim1 = headerBuffer.int
        val dim2 = headerBuffer.int
        val labelsPresent = headerBuffer.int  // 1 indicates labels are present

        // Calculate total number of floats
        val totalFloats = numArrays * dim1 * dim2
        println("numArrays: $numArrays, dim1: $dim1, dim2: $dim2, labelsPresent: $labelsPresent, totalFloats: $totalFloats")

        // Read data
        val dataSize = totalFloats * 4  // float32 is 4 bytes
        val dataBytes = ByteArray(dataSize)
        var bytesRead = 0
        while (bytesRead < dataSize) {
            val result = inputStream.read(dataBytes, bytesRead, dataSize - bytesRead)
            if (result == -1) break
            bytesRead += result
        }

        // Convert bytes to float array
        val dataBuffer = ByteBuffer.wrap(dataBytes).order(ByteOrder.LITTLE_ENDIAN)
        val floatBuffer = dataBuffer.asFloatBuffer()

        // Read labels if present
        val labels = IntArray(numArrays)
        if (labelsPresent == 1) {
            val labelsSize = numArrays * 4  // int32 is 4 bytes
            val labelsBytes = ByteArray(labelsSize)
            bytesRead = 0
            while (bytesRead < labelsSize) {
                val result = inputStream.read(labelsBytes, bytesRead, labelsSize - bytesRead)
                if (result == -1) break
                bytesRead += result
            }
            val labelsBuffer = ByteBuffer.wrap(labelsBytes).order(ByteOrder.LITTLE_ENDIAN)
            for (i in 0 until numArrays) {
                labels[i] = labelsBuffer.int
            }
        } else {
            // If labels are not present, you can handle accordingly
            // For this example, we'll set labels to -1
            for (i in 0 until numArrays) {
                labels[i] = -1
            }
        }

        inputStream.close()

        // Store data and labels in an array of DataSample objects
        val dataSamples = Array(numArrays) { index ->
            val sampleData = FloatArray(dim1 * dim2)
            floatBuffer.get(sampleData)
            DataSample(data = sampleData, label = labels[index])
        }

        return dataSamples
    }
}


